File: src/ru/hudspeed/pro/DrivingHudPresentation.java
package ru.gpsantiradar.app;

import java.util.Locale;

public final class DrivingHudPresentation {
    public static final int COLOR_GREEN = 0xFF00A652;
    public static final int COLOR_ALERT = 0xFFFFBE37;
    public static final int COLOR_OVERSPEED = 0xFFE53935;

    public final String speedText;
    public final String distanceText;
    public final String cameraText;
    public final int speedColor;
    public final boolean hasActiveObject;

    private DrivingHudPresentation(String speedText, String distanceText,
            String cameraText, int speedColor, boolean hasActiveObject) {
        this.speedText = speedText;
        this.distanceText = distanceText;
        this.cameraText = cameraText;
        this.speedColor = speedColor;
        this.hasActiveObject = hasActiveObject;
    }

    public static DrivingHudPresentation from(DrivingSnapshot snapshot) {
        return from(snapshot, AppSettings.DEFAULT_OVERSPEED_THRESHOLD_KMH);
    }

    public static DrivingHudPresentation from(DrivingSnapshot snapshot,
                                               int overspeedThresholdKmh) {
        boolean hasActiveObject = snapshot.hasObject();
        int speedColor = color(snapshot, hasActiveObject, overspeedThresholdKmh);
        String speedText = Integer.toString(Math.round(snapshot.speedKmh));
        String distanceText = snapshot.distanceMeters < 0 ? "—"
                : formatDistance(snapshot.distanceMeters);
        String cameraText = hasActiveObject
                ? snapshot.cameraName + limitSuffix(snapshot.speedLimitKmh)
                : "Объектов впереди нет";
        return new DrivingHudPresentation(speedText, distanceText, cameraText,
                speedColor, hasActiveObject);
    }

    private static int color(DrivingSnapshot snapshot, boolean hasActiveObject,
                             int overspeedThresholdKmh) {
        if (!hasActiveObject || snapshot.distanceMeters > snapshot.alertDistanceMeters
                || snapshot.speedLimitKmh <= 0
                || snapshot.speedKmh < snapshot.speedLimitKmh) {
            return COLOR_GREEN;
        }
        int threshold = AppSettings.clampOverspeedThreshold(overspeedThresholdKmh);
        return snapshot.speedKmh >= snapshot.speedLimitKmh + threshold
                ? COLOR_OVERSPEED : COLOR_ALERT;
    }

    private static String formatDistance(int meters) {
        return meters >= 1000
                ? String.format(Locale.US, "%.1f км", meters / 1000.0)
                : meters + " м";
    }

    private static String limitSuffix(int speedLimitKmh) {
        return speedLimitKmh > 0 ? "  ·  " + speedLimitKmh + " км/ч" : "";
    }
}

File: src/main/kotlin/ru/gpsantiradar/app/DrivingHudPresentation.kt
package ru.gpsantiradar.app

import java.util.Locale
import kotlin.math.roundToInt

class DrivingHudPresentation private constructor(
    val speedText: String,
    val distanceText: String,
    val cameraText: String,
    val speedColor: Int,
    val hasObject: Boolean,
    val hasActiveObject: Boolean
) {
    companion object {
        val COLOR_GREEN = 0xFF00A652.toInt()
        val COLOR_ALERT = 0xFFFFBE37.toInt()
        val COLOR_OVERSPEED = 0xFFE53935.toInt()

        fun from(snapshot: DrivingSnapshot) = from(snapshot, AppSettings.DEFAULT_OVERSPEED_THRESHOLD_KMH)

        fun from(snapshot: DrivingSnapshot, overspeedThresholdKmh: Int): DrivingHudPresentation {
            val hasObject = snapshot.hasObject()
            val hasActiveObject = snapshot.hasActiveCamera()
            val speedColor = color(snapshot, hasActiveObject, overspeedThresholdKmh)
            val distanceText = if (snapshot.distanceMeters < 0) "—" else formatDistance(snapshot.distanceMeters)
            val cameraText = if (hasObject) snapshot.cameraName + limitSuffix(snapshot.speedLimitKmh)
                else "Объектов впереди нет"
            return DrivingHudPresentation(
                snapshot.speedKmh.roundToInt().toString(), distanceText, cameraText, speedColor,
                hasObject, hasActiveObject,
            )
        }

        private fun color(snapshot: DrivingSnapshot, hasActiveObject: Boolean, overspeedThresholdKmh: Int): Int {
            if (!hasActiveObject || snapshot.speedLimitKmh <= 0 || snapshot.speedKmh < snapshot.speedLimitKmh
            ) return COLOR_GREEN
            val threshold = AppSettings.clampOverspeedThreshold(overspeedThresholdKmh)
            return if (snapshot.speedKmh >= snapshot.speedLimitKmh + threshold) COLOR_OVERSPEED else COLOR_ALERT
        }

        private fun formatDistance(meters: Int) =
            if (meters >= 1000) String.format(Locale.US, "%.1f км", meters / 1000.0) else "$meters м"
        private fun limitSuffix(speedLimitKmh: Int) = if (speedLimitKmh > 0) "  ·  $speedLimitKmh км/ч" else ""
    }
}
